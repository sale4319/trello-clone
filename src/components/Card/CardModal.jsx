import React, { useContext, useEffect, useState } from 'react';
import { RiWindow2Line, RiFileList2Line, RiDeleteBin6Fill, RiCloseFill } from 'react-icons/ri';
import { FiList } from 'react-icons/fi';

import { CardModalContext } from '../../providers/';
import {
    useEditCard,
    useDeleteCard,
    useSetCardDescription,
    useCreateComment,
    useEditCardComment,
    useDeleteCardComment,
    useLoadCardComments
} from '../../api/apiHooks/apiCards';

const CardModal = ({ card, handleClose }) => {
    const { setShowModal, selectedCardComments } = useContext(CardModalContext);

    const [cardName, setCardName] = useState(card.name);
    const [submitNewName, setSubmitNewName] = useState('');
    const [isEditingName, setIsEditingName] = useState(false);
    const { isSuccess: isEditNameSuccess } = useEditCard(submitNewName, card.id);

    const [description, setDescription] = useState(card.desc ?? '');
    const [submitDescription, setSubmitDescription] = useState('');
    const [isEditingDescription, setIsEditingDescription] = useState(false);
    const { isSuccess: isSetDescriptionSuccess } = useSetCardDescription(card.id, submitDescription);

    const [newComment, setNewComment] = useState('');
    const [submitNewComment, setSubmitNewComment] = useState('');
    const [isWritingComment, setIsWritingComment] = useState(false);
    useLoadCardComments(card.id);
    const { isSuccess: isCreateNewCommentSuccess } = useCreateComment(card.id, submitNewComment);

    const [isEditingComment, setIsEditingComment] = useState(false);
    const [editedComment, setEditedComment] = useState({ commentId: '', value: '' });
    const [submitCommentEditValue, setSubmitCommentEditValue] = useState('');
    const { isSuccess: isSuccessEditingComment } = useEditCardComment(
        card.id,
        editedComment.commentId,
        submitCommentEditValue,
    );

    const [submitDeleteCommentId, setSubmitDeleteCommentId] = useState('');
    const { isSuccess: isSuccessDeleteComment } = useDeleteCardComment(card.id, submitDeleteCommentId);

    const [submitDeleteCard, setSubmitDeleteCard] = useState('');
    const { isSuccess: isSuccessDeleteCard } = useDeleteCard(submitDeleteCard);

    useEffect(() => {
        return () => setShowModal(false);
    }, []);

    useEffect(() => {
        if (isCreateNewCommentSuccess && submitNewComment) {
            setSubmitNewComment('');
        }
    }, [isCreateNewCommentSuccess]);

    useEffect(() => {
        if (isSuccessDeleteComment && submitDeleteCommentId) {
            setSubmitDeleteCommentId('');
        }
    }, [isSuccessDeleteComment]);

    useEffect(() => {
        if (submitCommentEditValue && isSuccessEditingComment) {
            setSubmitCommentEditValue('');
        }
    }, [submitCommentEditValue]);

    useEffect(() => {
        if (isEditNameSuccess && submitNewName) {
            setSubmitNewName('');
        }
    }, [isEditNameSuccess]);

    useEffect(() => {
        if (isSetDescriptionSuccess && submitDescription) {
            setSubmitDescription('');
        }
    }, [isSetDescriptionSuccess]);

    useEffect(() => {
        if (isSuccessDeleteCard && submitDeleteCommentId) {
            setSubmitDeleteCard('');
        }
    }, [isSuccessDeleteCard]);

    const handleSaveComment = () => {
        if (newComment) {
            setSubmitNewComment(newComment);
            setNewComment('');
        }
        setIsWritingComment(false);
    };

    const handleOnClickEditComment = (comment) => {
        setIsEditingComment(true);
        setEditedComment({
            commentId: comment.id,
            value: comment.data.text,
        });
    };

    const handleDeleteComment = (commentId) => {
        setSubmitDeleteCommentId(commentId);
    };

    const handleDeleteCard = () => {
        setSubmitDeleteCard(card.id);
    };

    const handleKeyPressEnterEditComment = (event) => {
        if (event.key === 'Enter') {
            setIsEditingComment(false);
            setSubmitCommentEditValue(editedComment.value);
        }
    };

    const handleChangeEditComment = (event) => {
        setEditedComment({
            ...editedComment,
            value: event.target.value,
        });
    };

    const handleKeyPressEnterComment = (event) => {
        if (event.key === 'Enter') {
            handleSaveComment();
        }
    };

    const handleKeyPressEnterName = (event) => {
        if (event.key === 'Enter') {
            event.preventDefault();
            if (cardName && cardName !== card.name) {
                setSubmitNewName(cardName);
            }
            setIsEditingName(false);
        }
    };

    const handleBlurName = () => {
        if (cardName && cardName !== card.name) {
            setSubmitNewName(cardName);
        }
        setIsEditingName(false);
    };

    const handleDescriptionClick = () => setIsEditingDescription(true);

    const handleSaveDescription = () => {
        if (description !== card.desc) {
            setSubmitDescription(description);
        }
        setIsEditingDescription(false);
    };

    const handleCancelDescription = () => {
        setDescription(card.desc ?? '');
        setIsEditingDescription(false);
    };

    return (
        <div className="card-modal-container">
            <div className="card-modal-content">
                <div className="card-modal-body">
                    <div className="card-modal-title">
                        <RiWindow2Line className="card-modal-icon" />
                        <div className="card-modal-subtitle" onClick={() => setIsEditingName(true)}>
                            {!isEditingName ? (
                                <span>{cardName}</span>
                            ) : (
                                <input
                                    autoFocus
                                    value={cardName}
                                    onChange={event => setCardName(event.target.value)}
                                    onKeyPress={handleKeyPressEnterName}
                                    onBlur={handleBlurName}
                                />
                            )}
                        </div>
                        <RiDeleteBin6Fill
                            className="card-modal-delete-comment-icon flex-content-end"
                            size={25}
                            onClick={() => handleDeleteCard()}
                            title="Delete this card"
                        />
                    </div>
                    <div className="card-modal-divider" />

                    <div>
                        <div className="card-modal-title">
                            <RiFileList2Line className="card-modal-icon" />
                            <div className="card-modal-subtitle">Description</div>
                        </div>
                        <div>
                            <textarea
                                onClick={handleDescriptionClick}
                                value={description}
                                placeholder="Add a more detailed description"
                                className="card-modal-input"
                                onChange={event => setDescription(event.target.value)}
                                onBlur={() => setIsWritingComment(false)}
                            />
                        </div>
                        {isEditingDescription && (
                            <div className="card-modal-buttons-container">
                                <div className="card-modal-button card-modal-description-save-button card-modal-activity-button" onClick={handleSaveDescription}>
                                    SAVE
                                </div>
                                <div className="card-modal-description-close-button" onClick={handleCancelDescription}>
                                    <RiCloseFill />
                                </div>
                            </div>
                        )}
                    </div>

                    <div className="card-modal-divider" />

                    <div>
                        <div className="card-modal-title">
                            <FiList className="card-modal-icon" />
                            <span className="card-modal-subtitle">Activity</span>
                        </div>
                        <div>
                            <input
                                value={newComment}
                                placeholder="Write a comment..."
                                className="card-modal-activity-input"
                                type="text"
                                onClick={() => setIsWritingComment(true)}
                                onChange={event => setNewComment(event.target.value)}
                                onKeyPress={event => handleKeyPressEnterComment(event)}
                            />
                            <div
                                hidden={!isWritingComment || !newComment}
                                className="card-modal-button card-modal-description-save-button card-modal-activity-button"
                                onClick={handleSaveComment}>
                                Save
                            </div>

                            <div>
                                {selectedCardComments.map((comment) => (
                                    <div key={comment.id}>
                                        {isEditingComment && editedComment.commentId === comment.id ? (
                                            <div>
                                                <input
                                                    autoFocus
                                                    className="card-modal-edit-comment-input card-modal-editing-input"
                                                    value={editedComment.value}
                                                    onBlur={() => setIsEditingComment(false)}
                                                    onChange={handleChangeEditComment}
                                                    onKeyPress={handleKeyPressEnterEditComment}
                                                />
                                            </div>
                                        ) : (
                                            <div
                                                className="general-container">
                                                <div
                                                    className="card-modal-comment-content"
                                                    onClick={() => handleOnClickEditComment(comment)}>
                                                    {comment.data.text}
                                                </div>
                                                <div className="card-modal-comment-content-button">
                                                    <RiDeleteBin6Fill
                                                        className="card-modal-delete-comment-icon"
                                                        size={25}
                                                        onClick={() => handleDeleteComment(comment.id)}
                                                    />
                                                </div>
                                            </div>
                                        )}
                                    </div>
                                ))}

                            </div>
                        </div>
                    </div>

                    <div className="card-modal-divider" />

                    <div className="card-modal-footer">
                        <div className="card-modal-button card-modal-button-close" onClick={handleClose}>
                            Close
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
};

export default CardModal