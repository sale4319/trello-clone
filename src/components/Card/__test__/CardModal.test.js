import React from 'react';
import  CardModal  from '../CardModal';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

test('component renders and takes snapshot', () => {
    const component = render(<CardModal card={{name: "ss"}}></CardModal>);
    expect(component.container).toMatchSnapshot();
  });