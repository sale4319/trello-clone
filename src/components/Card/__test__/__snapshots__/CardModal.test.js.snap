// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders and takes snapshot 1`] = `
<div>
  <div
    class="card-modal-container"
  >
    <div
      class="card-modal-content"
    >
      <div
        class="card-modal-body"
      >
        <div
          class="card-modal-title"
        >
          <svg
            class="card-modal-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M3 3h18a1 1 0 0 1 1 1v16a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V4a1 1 0 0 1 1-1zm17 8H4v8h16v-8zm0-2V5H4v4h16zm-5-3h4v2h-4V6z"
              />
            </g>
          </svg>
          <div
            class="card-modal-subtitle"
          >
            <span>
              ss
            </span>
          </div>
          <svg
            class="card-modal-delete-comment-icon flex-content-end"
            fill="currentColor"
            height="25"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="25"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Delete this card
            </title>
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 4h5v2h-2v15a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1V6H2V4h5V2h10v2zM9 9v8h2V9H9zm4 0v8h2V9h-2z"
              />
            </g>
          </svg>
        </div>
        <div
          class="card-modal-divider"
        />
        <div>
          <div
            class="card-modal-title"
          >
            <svg
              class="card-modal-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M20 22H4a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h16a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1zm-1-2V4H5v16h14zM8 7h8v2H8V7zm0 4h8v2H8v-2zm0 4h5v2H8v-2z"
                />
              </g>
            </svg>
            <div
              class="card-modal-subtitle"
            >
              Description
            </div>
          </div>
          <div>
            <textarea
              class="card-modal-input"
              placeholder="Add a more detailed description"
            />
          </div>
        </div>
        <div
          class="card-modal-divider"
        />
        <div>
          <div
            class="card-modal-title"
          >
            <svg
              class="card-modal-icon"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="8"
                x2="21"
                y1="6"
                y2="6"
              />
              <line
                x1="8"
                x2="21"
                y1="12"
                y2="12"
              />
              <line
                x1="8"
                x2="21"
                y1="18"
                y2="18"
              />
              <line
                x1="3"
                x2="3.01"
                y1="6"
                y2="6"
              />
              <line
                x1="3"
                x2="3.01"
                y1="12"
                y2="12"
              />
              <line
                x1="3"
                x2="3.01"
                y1="18"
                y2="18"
              />
            </svg>
            <span
              class="card-modal-subtitle"
            >
              Activity
            </span>
          </div>
          <div>
            <input
              class="card-modal-activity-input"
              placeholder="Write a comment..."
              type="text"
              value=""
            />
            <div
              class="card-modal-button card-modal-description-save-button card-modal-activity-button"
              hidden=""
            >
              Save
            </div>
            <div />
          </div>
        </div>
        <div
          class="card-modal-divider"
        />
        <div
          class="card-modal-footer"
        >
          <div
            class="card-modal-button card-modal-button-close"
          >
            Close
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
